### ISSUE TITLE
store watch not triggered when depends on unregistered module

### ISSUE THREAD

**Issue by matthieusieben** [vuejs/vuex#1805]
### Version
3.5.1

### Reproduction link
[https://codesandbox.io/s/awesome-wescoff-97l47?file=/src/main.js](https://codesandbox.io/s/awesome-wescoff-97l47?file=/src/main.js)

<details>

<summary>Code</summary>

```js
import Vue from "vue";
import Vuex from "vuex";

Vue.use(Vuex);

// Uncomment one of these lines:
const initialState = '{"foo":{"value":4}}';
// const initialState = "{}";

const modules = {
  foo: {
    namespaced: true,
    getters: { value: state => state.value },
    state: () => ({ value: 1 })
  }
};

const myPlugin = store => {
  store.watch(
    (state, getters) => getters["foo/value"],
    value => console.error("Value from watch", value)
  );
};

const store = new Vuex.Store({
  state: () => JSON.parse(initialState),
  plugins: [myPlugin]
});

for (const ns in modules) {
  store.registerModule(ns, modules[ns], {
    preserveState: store.state[ns] !== undefined
  });
}

console.error("Value after init", store.getters["foo/value"]);
```

</details>

Notice the watcher's callback is not triggered.

The workaround is to register the watcher after the modules (i.e. not through a plugin)

### Steps to reproduce
- Create a store with 
  - a plugin that register a watcher for a module's getter
  - an initial state for the sub module
- Register the module (will generate a new store.getter object)

### What is expected?
The watcher should be triggered when the getter goes from "undefined" (before the module was registered) to "defined" (after the module was registered).

Basically, `store.getters` should be reactive.

### What is actually happening?
The watcher is not trigged

**Comment by kiaking** [comment]
Interesting use case, but this is not supported indeed. To watch to have an effect, you need to have key defined. In your case, the key is not `undefined`; there is no key in the first place.

We can't really make the whole getters watched, it's like watching entire Vue instance. I would have to suggest you to explicitly call `store.watch` function after registering the module.

**Comment by matthieusieben** [comment]
What bothers me here is that the watcher **does** work and **is** reactive when the store is _not_ pre populated, but **does not** work when the store _is_ pre populated (ssr).

**Comment by matthieusieben** [comment]
You can play with the following lines of code to see what I mean:
```js
// Uncomment one of these lines:
const initialState = '{"foo":{"value":4}}';
// const initialState = "{}";
```

**Comment by kiaking** [comment]
Oh I see, yeah well that makes more sense. Sorry for my misunderstanding. Though, it's because when the state is hydrated, nothing is really changing, hence watch doesn't fire. So again, there's not much we can do here.

Basically, watch was intended to actually watch the current state, and fire things up when ever it changes. This could be one of the downside having dynamic module registration 🤔
